[lammps] Adding new CMake package names

stephenmsachs opened the pull request:
This is a workaround that includes new and old package names. 

I currently do not see any ideal solution. You could look at every release which package names are the valid ones, but then the spack user would need to change their `variant` specifiers depending on Lammps version.

Another solution would be to have Spack translate package names. Then the question remains whether the old or new ones should be used as `variant` specifiers in Spack.

The full list of changes (with release dates) can be found here:
https://github.com/lammps/lammps/commits/patch_7Jan2022/cmake/presets/all_on.cmake

Commit:
Adding newer cmake package names

This is not really clean as old and new package names can be used. CMAKE will
ignore those not found, but Spack will not let the user know if the correct (old
or new) package name was used. The full list of changes (with release dates) can
be found here:
https://github.com/lammps/lammps/commits/patch_7Jan2022/cmake/presets/all_on.cmake

Commit:
Fill column to 88 chars

alalazo reviewed (changes requested):
> I currently do not see any ideal solution. You could look at every release which package names are the valid ones, but then the spack user would need to change their variant specifiers depending on Lammps version.

I think we can use [conditional variant values](https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-possible-values) to model this properly.

Commit:
adding whitespaces

Commit:
Divide packages in new/old/unchanged and adapt dependencies

Still need to conditionally add either `packages_pre_20210702` or `packages_post_20210702`.

stephenmsachs commented:
Hi @alalazo.  I don't think we can use conditional variants here as the packages need to be destinguishable from other variants when adding the to the CMake line. I picked all the updated package names apart and put them into three buckets (new/old/unchanged). I am having trouble extending `supported_packages` based on the version number (marked as TODO). Can you give me a hand here? I am sure this is a one-liner, but I just don't see it.

Commit:
Make flake8 happy

alalazo commented:
> Hi @alalazo. I don't think we can use conditional variants here as the packages need to be destinguishable from other variants when adding the to the CMake line. 

Can you elaborate on that (I am not super-familiar with lammps)? What is preventing us from e.g. using a single multi-valued variant and either:
1. Map a value in the variant to different CMake options depending on the version, or
2. Have conditional values exposed to users in the multi-valued variant
?

stephenmsachs commented:
> Can you elaborate on that (I am not super-familiar with lammps)? What is preventing us from e.g. using a single multi-valued variant and either:

I am not either. I just read through the `CmakeList.txt` and saw the changes in their github repo. We could use a single multi-valued variant but this would break existing lammps specs which use one variant per Cmake package.

> 
>     1. Map a value in the variant to different CMake options depending on the version, or
We could also do this, but (again not breaking existing specs) we would be stuck with the 'old' package names (e.g. `user-omp` instead of the now used `openmp`).
> 
>     2. Have conditional values exposed to users in the multi-valued variant
I don't fully understand this approach, but multi-valued variants would again break existing specs.

Commit:
Exposing CMake variables as conditioanl variants

Only the variant names that correspond to the Lammps version are valid in specs.

Commit:
[lammps] Remove duplicate `openmp` variant

Commit:
[lammps] flake8: line too long

alalazo commented:
@junghans @payerle This LGTM, but I'm definitely not an expert on lammps. Do you have further comments? Otherwise I'll go on merging the PR.

junghans commented:
Duplicate of #25015

rbberger commented:
@alalazo with #34011 now merged, should we close this one?
